# ２量子ビットの基本の操作を覚える
ここでは前回に引き続き２量子ビットの計算を覚えます。量子コンピュータでは２量子ビットの操作によってより複雑な計算を行うことができます。

In [0]:
!pip install blueqat

## 回路の作成
基本回路の作成方法は１量子ビットの場合と同じですが、今回は２量子ビットをつなげます。Blueqatでは最初に量子ビット数を指定する必要がありません（指定しても大丈夫です）。早速CXという回路を作ってみましょう。

In [3]:
from blueqat import Circuit

Circuit().cx[0,1].m[:].run(shots=1)

Counter({'00': 1})

上記回路では00という答えが出ました。回路の中にcx回路が入っていますので２量子ビットで答えが出ます。00は両方とも0という意味です。

## CX回路について
CX回路はCNOT回路とも呼ばれます。.cx[コントロールビット,ターゲットビット]とcxのゲートの中に「コントロールビット」と「ターゲットビット」と呼ばれる量子ビットを指定します。cxゲートはコントロールビットが1の時だけ、ターゲットビットを反転させます。


今回は両方とも量子ビットは0からスタートして、cxゲートがかかりました。0番目の量子ビットは0なので、1番目の量子ビットには何も操作がされません。

ちなみに回路は以下のようになります。

<img src="./img/002/002_0.png" width="25%">

## 再びCX回路
CX回路の効果を実感するために最初に0番目の量子ビットにXゲートを適用して0を1にしておきます。それからCX回路をかけてみましょう。


In [4]:
Circuit().x[0].cx[0,1].m[:].run(shots=1)

Counter({'11': 1})

今度は11という答えが出て両方が1になりました。これは、0番目の量子ビットがまずXゲートによって1になり、それに反応してCX回路が1番目の量子ビットを0から1に反転させたため、答えが11となっています。

## （応用）量子ビットの数を指定する
量子ビットの数を最初から指定することもできます。その際にはCircuit()のなかに引数を設定しましょう。

In [5]:
Circuit(3).x[0].cx[0,1].m[:].run(shots=1)

Counter({'110': 1})

110となりました。最初の2量子ビットは計算に使用されましたが、最後の量子ビットは使用されないままでした。以上です。

# 概要
ここでは２量子ビットの計算を説明します。

## 量子ビット
２量子ビットの場合は 00, 01, 10, 11 は以下のような状態ベクトルで表します。

<img src="./img/002/002_02_0.png" width="50%">

00 の状態ベクトルはテンソル積を用いると以下のように計算できます。

<img src="./img/002/002_02_1.png" width="50%">

## CX回路
CXゲートはコントロールビットが1ならばターゲットビットにXゲートを施すものです。

CXゲートは以下のような行列です。

<img src="./img/002/002_02_2.png" width="20%">

00 にCXゲートを施すと以下のようになります。

<img src="./img/002/002_02_3.png" width="43%">

この場合コントロールビットは0なので何も変化しません。

次に 10 にCXゲートを施すと以下のようになります。

<img src="./img/002/002_02_4.png" width="43%">

この場合コントロールビットは1なのでターゲットビットが反転して1になっているのがわかります。

## 観測
２量子ビットの場合の観測は 00, 01, 10, 11 が確率的に観測されます。   
それぞれ出力される確率は１量子ビットの場合と同じように各係数の絶対値の2乗となります。

例えば以下のような状態ベクトルを考えます。

<img src="./img/002/002_02_5.png" width="40%">

これを1回観測すると 00, 01, 10, 11 が出力される確率は 1/4 となります。

## 2量子ビット以上の場合
３量子ビットの場合で最初の２量子ビットにCXゲートをかけるとします。
100 にCXを施したとき計算は以下のようになります。

<img src="./img/002/002_02_6.png" width="45%">

上の式から最後のビットには何も変化しない行列Iを施して計算していることがわかります。